Fetch list of committers in a repository

pararthshah opened the pull request:
Added method Repository.Committers() to fetch the committers list in a git repo using `git shortlog -sne {rev}` command. The information returned for each committer consists of: name, email, # of commits in the specified branch/revision.

This method is only implemented for `gitcmd`; a mercurial implementation is pending.

shurcooL commented on vcs/gitcmd/repo.go:
```diff
@@ -27,6 +28,16 @@ import (
+const (
+	// LogEntryPattern is the regexp pattern that matches entries in the output of
+	// the `git shortlog -sne` command.
+	LogEntryPattern = `\s*([0-9]+)\s+([A-Za-z]+(?:\s[A-Za-z]+)*)\s+<([A-Za-z@.]+)>\s*`
```
Why does this need to be a part of API?

shurcooL commented on vcs/repository.go:
```diff
@@ -130,6 +133,14 @@ type CommitsOptions struct {
+type CommittersOptions struct {
+	N int // limit the number of returned committers, ordered by decreasing number of commits (0 means no limit)
+
+	Rev string // the rev for which committer stats will be fetched ("" means HEAD)
```
Does `HEAD` have meaning in general VCS terms, or is it git-specific?

Have you considered making this parameter mandatory, like `Head` is in `CommitsOptions`? Would that be a better or worse API?

shurcooL commented on vcs/gitcmd/repo.go:
```diff
@@ -890,6 +901,45 @@ func (r *Repository) Search(at vcs.CommitID, opt vcs.SearchOptions) ([]*vcs.Sear
+	if err != nil {
+		return nil, fmt.Errorf("exec `git shortlog -sne` failed: %v. Output was:\n\n%s", err, string(out))
+	}
+	out = bytes.TrimSpace(out)
```
You are using `cmd.CombinedOutput()` (which includes stdout and stderr) to get the value of `out` here, and then try to parse it.

There are two reasons to avoid doing that.

1. Stderr should not be necessary. The output you're looking to parse should be completely contained within `stdout`.

2. It can cause issues. Git (and other commands, in general) use stderr to output optional warning messages and the like. If those are present, it will likely cause parsing to fail or return invalid results. See issue #30 where it actually happened and related issue #33 that's still open.

I highly suggest using `cmd.Output()` to capture the value of stdout only, and parsing that. Stderr should not be needed at all, as it's only useful for reporting an error when it occurs. But I don't think it's important enough to spend extra effort capturing it.

pararthshah commented:
Addressed comments. I'm keeping CommittersOptions.Rev an optional arg since it is better to have the implementations handle that situation. The API clients can expect a reasonable committers list when they don't specify a revision. 'Reasonable' would mean the committers on the default/currently checked out revision. eg. the gitcmd implementation uses `HEAD`, while hgcmd could use `default`.

shurcooL commented:
Branch `03-hooks` has been merged into master; this PR can now be re-created with `master` as base branch. Closing since it shouldn't be merged into `03-hooks`.
